src/gDP.cpp

```diff
@@ -792,6 +797,147 @@ void gDPSetScissor(u32 mode, s16 xh, s16 yh, s16 xl, s16 yl)
+void gDPMemset(u32 value, u32 addr, u32 length)
+{
+	// TODO: Can this be different?
+	const u32 screenWidth = VI.width;
+	const u32 screenHeight = VI.height;
+
+	GraphicsDrawer& drawer = dwnd().getDrawer();
+	u32 fillColor = value | (value << 16);
+	const u32 depthImageStart = gDP.depthImageAddress;
+	const u32 depthImageEnd = gDP.depthImageAddress + screenWidth * screenHeight * 2;
+	if (depthImageStart <= addr && addr < depthImageEnd) {
+		// ZB clear
+		gDP.rectColor.r = _FIXED2FLOATCOLOR(_SHIFTR(fillColor, 11, 5), 5);
+		gDP.rectColor.g = _FIXED2FLOATCOLOR(_SHIFTR(fillColor, 6, 5), 5);
+		gDP.rectColor.b = _FIXED2FLOATCOLOR(_SHIFTR(fillColor, 1, 5), 5);
+		gDP.rectColor.a = _SHIFTR(fillColor, 0, 1);
+
+		// HACK: this usually replaces gDPSetColorImage for zb so save zb
+		frameBufferList().saveBuffer(gDP.depthImageAddress, (u16)G_IM_FMT_RGBA, (u16)G_IM_SIZ_16b, (u16)screenWidth, false);
+
+		if (config.generalEmulation.enableFragmentDepthWrite == 0)
+			drawer.clearDepthBuffer();
+		else
+			depthBufferList().setCleared(true);
+
+		// Deduce ulx, uly, lrx, lry from addr, size, and colorImageStart
+		// Currently I am assuming a simple letterbox case - top and bottom are not cleared
+		const u32 lineSize = screenWidth * 2;
+		const u32 ulx = 0;
+		const u32 uly = (addr - depthImageStart) / lineSize;
+		const u32 lrx = screenWidth;
+		const u32 lry = uly + length / lineSize;
+
+		if (config.generalEmulation.enableFragmentDepthWrite != 0) {
+			// Pretend that we are drawing the rectangle over ZB with fill mode
+			ValueKeeper<u32> backupColorImageAddress(gDP.colorImage.address, gDP.depthImageAddress);
+			ValueKeeper<u32> backupWidth(gDP.colorImage.width, screenWidth);
+			const unsigned int backupDepthSource = gDP.otherMode.depthSource;
+			const unsigned int backupCycleType = gDP.otherMode.cycleType;
+
+			gDP.otherMode.cycleType = G_CYC_FILL;
+
+			drawer.drawRect(ulx, uly, lrx, lry);
+			frameBufferList().setBufferChanged(f32(lry));
+
+			gDP.otherMode.depthSource = backupDepthSource;
+			gDP.otherMode.cycleType = backupCycleType;
+		}
+
+		// TODO: Following the same logic and not filling in zb values in RDRAM, why?
+		return;
+	}
+
+	// FB clear
+	bool doMemset = true;
+	if (0 == config.frameBufferEmulation.enable)
+	{
+		// Either RGBA16 or RGBA32. This heavily assume "niceness" of developers to bind color buffer before memset...
+		const u32 colorSize = gDP.colorImage.size < 3 ? 2 : 4;
+		const u32 colorImageStart = gDP.colorImage.address;
+		const u32 colorImageEnd = gDP.colorImage.address + ((screenWidth * screenHeight) << (gDP.colorImage.size >> 1));
+		if (colorImageStart <= addr && addr < colorImageEnd) {
+			f32 fillColor32[4];
+			getFillColor(fillColor, fillColor32);
+			gDP.rectColor.r = fillColor32[0];
+			gDP.rectColor.g = fillColor32[1];
+			gDP.rectColor.b = fillColor32[2];
+			gDP.rectColor.a = fillColor32[3];
+
+			const u32 lineSize = colorSize * screenWidth;
```

Comment:
wrong. should be 
```suggestion
const u32 lineSize = screenWidth << gDP.colorImage.size >> 1;
```
colorSize variable is useless.